### PR TITLE
Locations as groups

### PR DESCRIPTION
**Pull request by twymer**
This is kind of a monster, but the two main things that it accomplishes are:
- Starting to try to clean up the access/usage of user locations. It adds a feature flag to use to hide all of the multiple locations per user logic that IPM requires.
- Managing magic fake case sharing groups for locations. This was piggy backed on the previous feature because we thought it could replace delegate access to supply point cases, but we were wrong. Anyway, now case sharing exists, so whatever.

This PR is expected to contain all functionality of phase 1 and phase 2 from the spec doc (https://docs.google.com/a/dimagi.com/document/d/1C0dubLWtSFHr0CBzlrMOTbppmdBkfB7ip0DXiGVNBw0/) EXCEPT suite file changes. QA has been deferred for this PR, so this is not guaranteed.

### CODE REVIEW

**Comment by dimagimon** [comment]
:+1: No lint errors found.

File: corehq/apps/commtrack/helpers.py [hunk 1]
```diff
@@ -16,5 +16,5 @@ def make_product(domain, name, code):
     return p
 
 
-def make_supply_point(domain, location, owner_id=None):
-    return SupplyPointCase.create_from_location(domain, location, owner_id)
+def make_supply_point(domain, location):
+    return SupplyPointCase.create_from_location(domain, location)
```
Comment (twymer) [inline corehq/apps/commtrack/helpers.py]: It doesn't really make sense anymore to be able to set the owner, and this setting was only used in tests.

**Comment by czue** [comment]
@twymer are you looking for review on this?

**Comment by twymer** [comment]
@czue No, was just opening it to have the full build run since I can't do that locally.

File: corehq/apps/users/exceptions.py [hunk 2]
```diff
@@ -2,4 +2,7 @@ class NoAccountException(Exception):
     """
     Raised when trying to access the account of someone without one
     """
-    pass
\ No newline at end of file
+    pass
+
```
Comment (dimagimon) [inline corehq/apps/users/exceptions.py]: E302 expected 2 blank lines, found 1

File: corehq/apps/reports/filters/api.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 from braces.views import JSONResponseMixin
 
+from corehq.apps.domain.models import Domain
```
Comment (dimagimon) [inline corehq/apps/reports/filters/api.py]: F401 'Domain' imported but unused

File: corehq/apps/locations/_design/views/users_by_location_id/map.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+function(doc) {
```
Comment (dimagimon) [inline corehq/apps/locations/_design/views/users_by_location_id/map.js]: Missing name in function declaration.

File: corehq/apps/users/views/__init__.py [hunk 5]
```diff
@@ -190,7 +190,7 @@ def commtrack_form(self):
         user_domain_membership = self.editable_user.get_domain_membership(self.domain)
         linked_loc = user_domain_membership.location_id
         linked_prog = user_domain_membership.program_id
-        return CommtrackUserForm(domain=self.domain, is_admin=self.request.couch_user.is_domain_admin(self.domain), initial={'supply_point': linked_loc, 'program_id': linked_prog})
+        return CommtrackUserForm(domain=self.domain, is_admin=self.request.couch_user.is_domain_admin(self.domain), initial={'location': linked_loc, 'program_id': linked_prog})
```
Comment (dimagimon) [inline corehq/apps/users/views/__init__.py]: E501 line too long (176 > 115 characters)

File: corehq/apps/users/bulkupload.py [hunk 6]
```diff
@@ -17,6 +17,7 @@
 from corehq.apps.custom_data_fields import CustomDataFieldsDefinition
 from corehq.apps.groups.models import Group
 from corehq.apps.domain.models import Domain
+from corehq.apps.locations.models import SQLLocation, LOCATION_SHARING_PREFIX
```
Comment (dimagimon) [inline corehq/apps/users/bulkupload.py]: F401 'LOCATION_SHARING_PREFIX' imported but unused

File: corehq/apps/locations/tests/test_locations.py [hunk 7]
```diff
@@ -10,7 +11,36 @@
...
+                administrative=True
+            ),
+            LocationType(
+                name='outlet',
+                allowed_parents=['village']
+            ),
+        ]
+        self.domain.save()
+
+        self.loc = make_loc('loc', type='outlet',  domain=self.domain.name)
```
Comment (dimagimon) [inline corehq/apps/locations/tests/test_locations.py]: E241 multiple spaces after ','

**Comment by twymer** [comment]
test

File: corehq/apps/locations/models.py [hunk 8]
```diff
@@ -12,6 +12,10 @@
 from corehq.apps.domain.models import Domain
 from corehq.apps.products.models import SQLProduct
 from mptt.models import MPTTModel, TreeForeignKey
+from corehq.apps.domain.models import Domain
```
Comment (dimagimon) [inline corehq/apps/locations/models.py]: F811 redefinition of unused 'Domain' from line 12

File: corehq/apps/commtrack/views.py [hunk 9]
```diff
@@ -180,7 +180,7 @@ def loc_to_payload(loc):
 
     else:
         LIMIT = 100
-        loc_types = [loc_type.name for loc_type in Domain.get_by_name(domain).location_types if not loc_type.administrative]
+        loc_types = [loc_type.name for loc_type in Domain.get_by_name(domain).location_types]
```
Comment (czue) [inline corehq/apps/commtrack/views.py]: since we can now assign users to administrative locations we need to stop filtering the list.

File: corehq/apps/fixtures/models.py [hunk 10]
```diff
@@ -287,14 +288,37 @@ def _serialize(val):
...
+                    groups.append(loc.reporting_group_object())
+
+            return set(
+                list(Group.view(
+                    '_all_docs',
+                    keys=list(group_ids.difference(location_group_ids)),
+                    include_docs=True
+                )) +
+                groups
+            )
```
Comment (twymer) [inline corehq/apps/fixtures/models.py]: write some comments..

File: corehq/apps/fixtures/models.py [hunk 11]
```diff
@@ -287,14 +288,37 @@ def _serialize(val):
...
                 reduce=False,
                 wrapper=lambda r: r['value']
             )
         )
+
         if wrap:
-            return set(Group.view('_all_docs', keys=list(group_ids), include_docs=True))
+            location_group_ids = set([
+                gid for gid in group_ids
+                if gid.startswith('locationgroup-') or gid.startswith('locationreportinggroup-')
```
Comment (czue) [inline corehq/apps/fixtures/models.py]: should use `LOCATION_SHARING_PREFIX` constants

File: corehq/apps/fixtures/models.py [hunk 12]
```diff
@@ -287,14 +288,37 @@ def _serialize(val):
...
+                'fixtures/ownership',
                 key=[self.domain, 'group by data_item', self.get_id],
                 reduce=False,
                 wrapper=lambda r: r['value']
             )
         )
+
         if wrap:
-            return set(Group.view('_all_docs', keys=list(group_ids), include_docs=True))
+            location_group_ids = set([
```
Comment (czue) [inline corehq/apps/fixtures/models.py]: should add comment "if any fixtures are referencing location group IDs, make sure that those get wrapped properly as group-looking things"

File: corehq/apps/importer/util.py [hunk 13]
```diff
@@ -346,15 +347,37 @@ def get_spreadsheet(download_ref, column_headers=True):
         return None
     return ExcelFile(download_ref.get_filename(), column_headers)
 
+
+def is_location_group(owner_id, domain):
+    """
+    Return yes if the specified owner_id is one of the
+    faked location groups.
+    """
+    group_ids = [
```
Comment (czue) [inline corehq/apps/importer/util.py]: can this be replaced with `owner_id.startswith(PREFIX)`?

Comment (czue) [inline corehq/apps/importer/util.py]: if not - it at least seems like you could pull out the id generation here and call it instead of referencing the object's ID.

File: corehq/apps/locations/_design/views/users_by_location_id/map.js (hunk 4 above)
Comment (czue) [inline corehq/apps/locations/_design/views/users_by_location_id/map.js]: i think we could use elastic for this - though if this is called in an OTA restore workflow then we should just bite the bullet and add a couch view.

File: corehq/apps/locations/fixtures.py [hunk 14]
```diff
@@ -58,7 +58,16 @@ def location_fixture_generator(user, version, last_sync=None):
     if toggles.SYNC_ALL_LOCATIONS.enabled(user.domain):
         location_db = _location_footprint(Location.by_domain(user.domain))
     else:
-        location_db = _location_footprint(user.locations)
+        locations = []
```
Comment (czue) [inline corehq/apps/locations/fixtures.py]: this adds the descendants to the fixture. might be worth putting a comment in.

File: corehq/apps/locations/management/commands/set_location_id.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
+from django.core.management.base import BaseCommand
+from dimagi.utils.couch.database import iter_docs
+from corehq.apps.users.models import CouchUser, CommCareUser
+
+
+class Command(BaseCommand):
+    help = ''
+
+    def handle(self, *args, **options):
```
Comment (czue) [inline corehq/apps/locations/management/commands/set_location_id.py]: this needs to be run after deploy to set the location id instead of dropping the ID in a list

File: corehq/apps/locations/models.py [hunk 16]
```diff
@@ -85,6 +89,84 @@ def root_locations(cls, domain, include_archive_ancestors=False):
         roots = cls.objects.root_nodes().filter(domain=domain)
         return _filter_for_archived(roots, include_archive_ancestors)
 
+    def case_sharing_group_object(self, user_id=None):
+        """
+        Returns a fake group object that SHOULD NOT be saved.
```
Comment (czue) [inline corehq/apps/locations/models.py]: can we make a subclass of `Group` that can enforce this by overriding save?

File: corehq/apps/locations/models.py [hunk 17]
```diff
@@ -85,6 +89,84 @@ def root_locations(cls, domain, include_archive_ancestors=False):
...
+
+        return g
+
+    def reporting_group_object(self, user_id=None):
+        """
+        Returns a fake group object that SHOULD NOT be saved.
+
+        Similar to case_sharing_group_object method, but for
+        reporting groups.
+        """
```
Comment (czue) [inline corehq/apps/locations/models.py]: can these both call `_make_group(user_id, name, case_sharing, reporting)` or similar?

File: corehq/apps/locations/models.py [hunk 18]
```diff
@@ -85,6 +89,84 @@ def root_locations(cls, domain, include_archive_ancestors=False):
...
+        g.metadata = {
+            'commcare_location_type': self.location_type,
+            'commcare_location_name': self.name,
+        }
+        for key, val in self.metadata.items():
+            g.metadata['commcare_location_' + key] = val
+
+        return g
+
+    def couch_location(self):
```
Comment (czue) [inline corehq/apps/locations/models.py]: should this be memoized?

File: corehq/apps/locations/models.py [hunk 19]
```diff
@@ -85,6 +89,84 @@ def root_locations(cls, domain, include_archive_ancestors=False):
...
+        return Location.get(self.location_id)
+
+    @memoized
+    def get_products(self):
```
Comment (twymer) [inline corehq/apps/locations/models.py]: ... why is this here?

File: corehq/apps/locations/models.py [hunk 20]
```diff
@@ -85,6 +89,84 @@ def root_locations(cls, domain, include_archive_ancestors=False):
...
+            )
+
+        g = Group()
+        g.domain = self.domain
+        g.name = group_name()
+        g.users = [user_id] if user_id else []
+        g.case_sharing = False
+        g.reporting = True
+        g.last_modified = datetime.now()
+        g._id = LOCATION_REPORTING_PREFIX + self.location_id
```
Comment (czue) [inline corehq/apps/locations/models.py]: might be worth abstracting this to `_get_location_group_id(location_id)`

File: corehq/apps/locations/models.py [hunk 21]
```diff
@@ -408,10 +490,22 @@ def linked_supply_point(self):
...
     @property
+    def group_id(self):
+        """
+        Returns the id with a prefix because this is
+        the magic id we are force setting the locations
+        case sharing group to be.
+
+        This is also the id that owns supply point cases.
+        """
+        return LOCATION_SHARING_PREFIX + self._id
```
Comment (czue) [inline corehq/apps/locations/models.py]: `_get_location_group_id(self._id)`

File: corehq/apps/locations/tests/test_locations.py [hunk 22]
```diff
@@ -210,6 +251,130 @@ def compare(list1, list2):
...
+        # reporting group is similar but has no ending
+        self.assertEqual(
+            'teststate/testvillage/testoutlet',
+            self.test_outlet.sql_location.reporting_group_object().name
+        )
+
+    def test_id_assignment(self):
+        # each should have the same id, but with a different prefix
+        self.assertEqual(
+            'locationgroup-' + self.test_outlet._id,
```
Comment (czue) [inline corehq/apps/locations/tests/test_locations.py]: use constants?

File: corehq/apps/locations/tests/test_locations.py [hunk 23]
```diff
@@ -210,6 +251,130 @@ def compare(list1, list2):
...
+        # group with a special prefix
+        self.loc.metadata = {
+            'foo': 'bar',
+            'fruit': 'banana'
+        }
+        self.loc.save()
+
+        self.assertDictEqual(
+            {
+                'commcare_location_type': self.loc.location_type,
```
Comment (czue) [inline corehq/apps/locations/tests/test_locations.py]: do we want to remove this prefix? i don't think there's any other way to get key/values into here

File: corehq/apps/reports/filters/users.py [hunk 24]
```diff
@@ -241,6 +241,32 @@ def user_types(self):
...
         return user_type_tuples
 
+    def get_location_groups(self):
+        def case_share_types():
+            return [
+                loc_type for loc_type in Domain.get_by_name(self.domain).location_types
+                if loc_type.shares_cases
+            ]
+
+        from corehq.apps.commtrack.models import SQLLocation
```
Comment (czue) [inline corehq/apps/reports/filters/users.py]: does this need to be here instead of at the top of the file?

File: corehq/apps/reports/filters/users.py [hunk 25]
```diff
@@ -296,6 +322,20 @@ def selected_sharing_group_ids(cls, request):
         emws = request.GET.getlist(cls.slug)
         return [g[4:] for g in emws if g.startswith("sg__")]
 
+    @classmethod
+    def selected_location_sharing_group_ids(cls, request):
+        emws = request.GET.getlist(cls.slug)
+        return [
+            g for g in emws if g.startswith('locationgroup-')
```
Comment (czue) [inline corehq/apps/reports/filters/users.py]: use constants

File: corehq/apps/reports/filters/users.py [hunk 26]
```diff
@@ -296,6 +322,20 @@ def selected_sharing_group_ids(cls, request):
...
+        ]
+
+    @classmethod
+    def selected_location_reporting_group_ids(cls, request):
+        emws = request.GET.getlist(cls.slug)
+        return [
+            g for g in emws if g.startswith('locationreportinggroup-')
```
Comment (czue) [inline corehq/apps/reports/filters/users.py]: .

File: corehq/apps/reports/standard/cases/basic.py [hunk 27]
```diff
@@ -157,17 +157,55 @@ def case_owners(self):
...
+        locations or those locations descendants.
+        """
+        from corehq.apps.locations.models import SQLLocation
+        from corehq.apps.users.models import CommCareUser
+        results = []
+        selected_location_group_ids = EMWF.selected_location_reporting_group_ids(self.request)
+
+        for group_id in selected_location_group_ids:
+            loc = SQLLocation.objects.get(
+                location_id=group_id.replace('locationreportinggroup-', '')
```
Comment (czue) [inline corehq/apps/reports/standard/cases/basic.py]: .

File: corehq/apps/users/bulkupload.py [hunk 28]
```diff
@@ -372,6 +426,13 @@ def is_password(password):
                     if email:
                         user.email = email
                     user.save()
+                    if location_code:
+                        loc = location_cache.get(location_code)
+                        if user.location_id != loc._id:
+                            # this triggers a second user save so
+                            # we want to avoid doing it if it isn't
+                            # needed
+                            user.set_location(loc)
```
Comment (twymer) [inline corehq/apps/users/bulkupload.py]: avoid saving user twice

File: corehq/apps/users/forms.py [hunk 29]
```diff
@@ -416,17 +418,25 @@ def __init__(self, *args, **kwargs):
...
 
             # add the supply point case id to user data fields
             # so that the phone can auto select
             supply_point = SupplyPointCase.get_by_location(loc)
-            user.user_data['commtrack-supply-point'] = supply_point._id
+            if supply_point:
+                user.user_data['commtrack-supply-point'] = supply_point._id
+
+            user.user_data['commcare_primary_case_sharing_id'] = \
+                LOCATION_SHARING_PREFIX + location_id
```
Comment (twymer) [inline corehq/apps/users/forms.py]: Only set this if it is a case sharing location type.

File: corehq/apps/users/models.py [hunk 30]
```diff
@@ -1310,46 +1312,101 @@ def fn(domain=None, data=None):
...
+            # the bulk upload?
+
+            # we only add the new one because we don't know
+            # if we can actually remove the old..
+            self.add_location(location)
         else:
-            return None
+            self.create_location_delegates([location])
+
+        self.user_data.update({
```
Comment (twymer) [inline corehq/apps/users/models.py]: only update if this type shares cases

File: corehq/apps/users/models.py [hunk 31]
```diff
@@ -1310,46 +1312,101 @@ def fn(domain=None, data=None):
...
+            # TODO is it possible to only remove this
+            # access if it was not previously granted by
+            # the bulk upload?
+
+            # we only add the new one because we don't know
+            # if we can actually remove the old..
+            self.add_location(location)
         else:
-            return None
+            self.create_location_delegates([location])
```
Comment (czue) [inline corehq/apps/users/models.py]: should only do this for non-administrative locations

File: corehq/apps/users/models.py (hunk 30 above)
Comment (czue) [inline corehq/apps/users/models.py]: and this one too

File: corehq/apps/users/models.py [hunk 32]
```diff
@@ -1365,53 +1422,74 @@ def supply_point_index_mapping(self, supply_point, clear=False):
...
         if mapping:
             mapping.delete()
 
-    def set_locations(self, locations):
+    def create_location_delegates(self, locations):
+        """
+        Submit the case blocks creating the delgate case access
+        for the location(s).
+        """
         from corehq.apps.commtrack.models import SupplyPointCase
```
Comment (czue) [inline corehq/apps/users/models.py]: you could add an assert here saying that type should not be `administrative_only`

Comment (czue) [inline corehq/apps/users/models.py]: nvm check is below

**Comment by czue** [comment]
(build is failing)

File: corehq/apps/importer/util.py (hunk 13 above)
Comment (twymer) [inline corehq/apps/importer/util.py]: For posterity: Couldn't just replace it as this check also handles verifying the group is for a location in this domain, but I refactored.

File: corehq/apps/users/bulkupload.py (hunk 28 above)
Comment (twymer) [inline corehq/apps/users/bulkupload.py]: Ugh no this was on purpose. When setting that delegate case, the user needs to be a saved user. So doing only one save would crash on creation.

Comment (czue) [inline corehq/apps/users/bulkupload.py]: makes sense

**Comment by czue** [comment]
:tada: and :pray:
